akka-actor/src/main/scala/akka/pattern/BackoffOptions.scala

```diff
@@ -71,14 +71,17 @@ object Backoff {
    * @param randomFactor after calculation of the exponential back-off an additional
    *   random delay based on this factor is added, e.g. `0.2` adds up to `20%` delay.
    *   In order to skip this additional delay pass in `0`.
+   * @param maxRestartAttempts maximum number of attempts to restart the child actor.
+   *   In order to restart infinitely pass in `-1`
    */
   def onFailure(
-    childProps:   Props,
-    childName:    String,
-    minBackoff:   FiniteDuration,
-    maxBackoff:   FiniteDuration,
-    randomFactor: Double): BackoffOptions =
-    BackoffOptionsImpl(RestartImpliesFailure, childProps, childName, minBackoff, maxBackoff, randomFactor)
+    childProps:         Props,
+    childName:          String,
+    minBackoff:         FiniteDuration,
+    maxBackoff:         FiniteDuration,
+    randomFactor:       Double,
+    maxRestartAttempts: Int): BackoffOptions =
```

Comment:
Unfortunately 'just' adding an additional parameter here will break backwards compatibility, so you'd need to add an extra (deprecated) `onFailure` method with the old signature.